go/tasks/plugins/webapi/databricks/plugin.go

```diff
@@ -0,0 +1,316 @@
+	"github.com/flyteorg/flytestdlib/errors"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/flyteorg/flytestdlib/promutils"
+
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery"
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/webapi"
+)
+
+const (
+	ErrSystem errors.ErrorCode = "System"
+	post      string           = "POST"
+	get       string           = "GET"
+)
+
+// for mocking/testing purposes, and we'll override this method
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type Plugin struct {
+	metricScope promutils.Scope
+	cfg         *Config
+	client      HTTPClient
+}
+
+type ResourceWrapper struct {
+	StatusCode     int
+	LifeCycleState string
+	JobID          string
+	Message        string
+}
+
+type ResourceMetaWrapper struct {
+	RunID              string
+	DatabricksInstance string
+	Token              string
+}
+
+func (p Plugin) GetConfig() webapi.PluginConfig {
+	return GetConfig().WebAPI
+}
+
+func (p Plugin) ResourceRequirements(_ context.Context, _ webapi.TaskExecutionContextReader) (
+	namespace core.ResourceNamespace, constraints core.ResourceConstraintsSpec, err error) {
+
+	// Resource requirements are assumed to be the same.
+	return "default", p.cfg.ResourceConstraints, nil
+}
+
+func (p Plugin) Create(ctx context.Context, taskCtx webapi.TaskExecutionContextReader) (webapi.ResourceMeta,
+	webapi.Resource, error) {
+	taskTemplate, err := taskCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	token, err := taskCtx.SecretManager().Get(ctx, p.cfg.TokenKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	container := taskTemplate.GetContainer()
+	sparkJob := plugins.SparkJob{}
+	err = utils.UnmarshalStruct(taskTemplate.GetCustom(), &sparkJob)
+	if err != nil {
+		return nil, nil, errors.Wrapf(pluginErrors.BadTaskSpecification, err, "invalid TaskSpecification [%v], failed to unmarshal", taskTemplate.GetCustom())
+	}
+
+	modifiedArgs, err := template.Render(ctx, container.GetArgs(), template.Parameters{
+		TaskExecMetadata: taskCtx.TaskExecutionMetadata(),
+		Inputs:           taskCtx.InputReader(),
+		OutputPath:       taskCtx.OutputWriter(),
+		Task:             taskCtx.TaskReader(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	decodeBytes, err := base64.StdEncoding.DecodeString(sparkJob.DatabricksConf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode databricksJob: %v: %v", sparkJob.DatabricksConf, err)
+	}
+
+	databricksJob := make(map[string]interface{})
+	err = json.Unmarshal(decodeBytes, &databricksJob)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal databricksJob: %v: %v", decodeBytes, err)
+	}
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	databricksJob["spark_python_task"] = map[string]interface{}{"python_file": p.cfg.EntrypointFile, "parameters": modifiedArgs}
```

Comment:
Does this assume, that all task code will be in one file or do you push it into master and workers? Also does this allow for the team to write packages?

Reply:
I wrote a entrypoint file like below. basically, we run this file first when running a spark job. it will execute the command we pass in the spark `parameters` (pyflyte-execute .....).

The reason to do that is because we need to run [pyflyte-fast-execute](https://github.com/flyteorg/flytekit/blob/2340e5dbaaea75f7b0c0418a92ace654b7bcf6d7/flytekit/bin/entrypoint.py) to execute some flytekit code. DBX doesn't allow us to override the default command, which always is `python spark_job.py`

What `pyflyte-fast-execute` does are
1. Download the workflow code (a zip file) to the driver. The code is uploaded when we register the workflow.
2. Download the task input (a proto file) from S3.
3. Deserialize proto to python value
4. Run the spark task
5. Serialize the output to proto
6. Upload the output to S3

```python
import os
import subprocess
import sys


def main():
    args = sys.argv
    print(args[1:])
    p = subprocess.run(args[1:], capture_output=True)
    print("==========================env==============================")
    print("AWS_ACCESS_KEY_ID", os.environ.get("AWS_ACCESS_KEY_ID"))
    print("==========================stdout==============================")
    print(p.stdout.decode('utf-8'))
    print("==========================stderr==============================")
    print(p.stderr.decode('utf-8'))


if __name__ == '__main__':
    main()

```